File: go/wrapper/moq/cache.go
package moq

import ffi "github.com/moq-dev/moq-go-ffi/moq"

// Cache is a shared, cheaply cloneable handle to a RAM LRU group cache.
//
// Attach it to a broadcast (or origin) with WithCache; pass the same Cache to several to pool one
// retention budget across them. Eviction is LRU by wall-clock last-access time, bounded by a byte
// budget and an age window.
type Cache struct {
	inner *ffi.MoqCache
}

// NewCache creates a cache with the given configuration. Use the zero CacheConfig for the default
// 64 MiB / 5s budget.
func NewCache(config CacheConfig) *Cache {
	return &Cache{inner: ffi.NewMoqCache(config)}
}

// CloneHandle returns a handle sharing this cache's budget. Attach it elsewhere to pool retention.
func (c *Cache) CloneHandle() *Cache {
	return &Cache{inner: c.inner.CloneHandle()}
}

// IsClone reports whether two handles share the same underlying budget.
func (c *Cache) IsClone(other *Cache) bool {
	return c.inner.IsClone(other.inner)
}

File: swift/Sources/Moq/Cache.swift
import MoqFFI

/// A shared, cheaply cloneable handle to a RAM LRU group cache.
///
/// Attach it to a broadcast (or origin) with `withCache`; pass the same `Cache` to several to
/// pool one retention budget across them. Eviction is LRU by wall-clock last-access time, bounded
/// by a byte budget and an age window.
public final class Cache: Sendable {
    let ffi: MoqCache

    /// Create a cache with the given configuration (defaults to a 64 MiB / 5s budget).
    public init(config: CacheConfig = CacheConfig()) {
        ffi = MoqCache(config: config)
    }

    init(_ ffi: MoqCache) {
        self.ffi = ffi
    }

    /// A handle sharing this cache's budget. Attach it elsewhere to pool retention.
    public func cloneHandle() -> Cache {
        Cache(ffi.cloneHandle())
    }

    /// Whether two handles share the same underlying budget.
    public func isClone(_ other: Cache) -> Bool {
        ffi.isClone(other: other.ffi)
    }
}
